### PR TITLE
How to build an NLP based emotion detection model using NeatText and Scikit-learn

### LINKED ISSUE #4433: [Machine Learning] How to build an NLP based emotion detection model using NeatText and Scikit-learn
#### Proposed title of article
[Machine Learning] How to build an NLP based emotion detection model using NeatText and Scikit-learn

#### Proposed article introduction

Natural language processing helps computers to understand and understand speech and written text like a human being. For example, NLP makes it possible for computers to read and analyze text, hear speech, interpret it and measure sentiment.

The newest type of NLP application is emotion text in text. The emotion detection model is a type of model that is used to detect the type of feeling in a given text. It may be a feeling of anger, joy, or happiness.

This emotion detection model can be used by different businesses to know if the customers are satisfied with their products. This helps the businesses to know if they can improve the quality of their product.

In this tutorial, we will use NeatText and Scikit-learn in building our model. NeatText is a Python library that is used to preprocess our dataset. NeatText will be used to clean the text dataset and remove stop words, these makes it easy for the model to use this dataset.

We then use Scikit-learn in building our model, it contains all the algorithms required for classification. This is a practical guide from data preprocessing to model building and testing.

#### Key takeaways 
1. What is mention detection.
2. How to install and use NeatText.
3. Dataset preprocessing.
4. Building the emotion detection model using Scikit-learn.
5. Testing the model.
#### Article quality
We will build the emotion detection model, this is a very crucial type of sentiment analysis. That's why in this tutorial we will start from scratch so that any user can easily follow along. This being a custom model it will be easy to explain key concepts about emotion detection.
This article is also unique because we use NeatTecx, for dataset preprocessing. This ensures we have a clean dataset to be used by the model. This tutorial follows all the stages from data preprocessing to model building and testing.

By the end of the tutorial, the model build can be applied to any business and help to solve real-world problems. This helps businesses to scale and know what adjustments they should make on their products to give them a competitive advantage.

### Conclusion
Finally, remove the Pre-Submission advice section and all our blockquoted notes as you fill in the form before you submit.
We look forwarding to reviewing your topic suggestion.

### Templates to use as guides
- [How To Guide Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/how-to-guide-template.md)
- [Software Review Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/software-review-template.md)
- [Tutorial Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/tutorial-template.md)

### PR DESCRIPTION
**Pull request by francis966**
#### Formatting and structure checks
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**

- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Is the article you are submitting an in-depth and unique article? Does it go beyond what is in the official docs and what is covered in other blog sites. See these articles as [examples](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md).
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**

- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) - [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker. We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**
- [x] Have you checked your code runs correctly and you've highlighted all necessary dependencies for installation?
- [x] Are the software programs and packages you're highlighting in your article up to date, using current versions and not deprecated?

This closes #4433

### CODE REVIEW

**Comment by ahmadmardeni1** [comment]
Please always use the template provided by us.
I edited your initial comment and added the template. Check what matches your article.

**Comment by Bayler** [comment]
Hi @francis966 Please remove the white spaces in your tutorial. Only add them when necessary i.e., when adding code snippets. Otherwise, your tutorial formatting looks great!

**Comment by francis966** [comment]
Hello @Bayler, I have removed the whitespaces where necessary.
Thanks

**Comment by ahmadmardeni1** [comment]
Plagiarism check done ✅

File: content/articles/how-to-build-an-nlp-based-emotion-detection-model-using-neattext-and-scikit-learn/index.md [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+Let's now start cleaning our dataset using Neattext.
+
+### Getting started with Neattext
+As mentioned earlier, Neattext is a Python library that is used to pre-process our dataset. Neattext will clean the text dataset by removing stop words and other noise.
+
+To install Neattext, run this command.
+
+```python
+!pip install neattext
+```
```
Comment (Bayler) [inline content/articles/how-to-build-an-nlp-based-emotion-detection-model-using-neattext-and-scikit-learn/index.md:102]: This ought to be a bash command. Not a python command

File: content/articles/how-to-build-an-nlp-based-emotion-detection-model-using-neattext-and-scikit-learn/index.md [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+#### Removing user handles
+The dataset contains some Twitter handles of different users. This is noise to our dataset, and it's unnecessary. We use`remove_userhandles` to remove them from our dataset.
+
+```python
+df['Clean_Text'] = df['Text'].apply(nfx.remove_userhandles)
+```
+We use `apply()` to add `remove_userhandles`. We save the cleaned dataset into a new column named `Clean_Text`.
+
+#### Removing stopwords
+Stopwords is a list of all the commonly used words in any language. The words being common they have little classification power. These words may lead to model bias during training.
```
Comment (Bayler) [inline content/articles/how-to-build-an-nlp-based-emotion-detection-model-using-neattext-and-scikit-learn/index.md]: What does this sentence, "The words being common they have little classification power." mean?

File: content/articles/how-to-build-an-nlp-based-emotion-detection-model-using-neattext-and-scikit-learn/index.md [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+#### CountVectorizer
+Machine learning models have a problem comprehending raw text. Machines cannot process the raw text data and it has to be converted into a matrix of numbers. Machines work well with numbers.
+
+CountVectorizer is used to convert the raw text into a matrix of numbers. This process depends on the frequency of each word in the entire text. During this process, `CountVectorizer` extracts important features from the text. They are then used as input for the model during training.
+
+#### train_test_split
+This method is important during the splitting of the dataset. It splits the dataset set into two sets, a train set, and a test set. This depends on the percentage specified by the user.
+
+#### accuracy_score
+This is important when calculating the accuracy score of our model during prediction.
```
Comment (Bayler) [inline content/articles/how-to-build-an-nlp-based-emotion-detection-model-using-neattext-and-scikit-learn/index.md:168]:
```suggestion
- `LogisticRegression` is an algorithm used for both classification and regression. This algorithm is imported from [Scikit-learn](https://scikit-learn.org/stable/). It will be used for emotion classification.

- Machine learning models have a problem comprehending raw text. Machines cannot process the raw text data and it has to be converted into a matrix of numbers. Machines work well with numbers. `CountVectorizer` is used to convert the raw text into a matrix of numbers. This process depends on the frequency of each word in the entire text. During this process, `CountVectorizer` extracts important features from the text. They are then used as input for the model during training.

- The `train_test_split` method is important during the splitting of the dataset. It splits the dataset set into two sets, a train set, and a test set. This depends on the percentage specified by the user.

- The `accuracy_score` is important when calculating the accuracy score of our model during prediction.
```

File: content/articles/how-to-build-an-nlp-based-emotion-detection-model-using-neattext-and-scikit-learn/index.md [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+Let's import the `Pipeline` using the following code.
+
+```python
+from sklearn.pipeline import Pipeline
+```
+To use `Pipeline` we need to specify the machine learning stages we want to automate. In this tutorial, we have two processes we want to automate.
+
+The first stage is the `CountVectorizer` process. This stage is used to convert the raw text dataset into a matrix of numbers that a machine can understand.
+
+The second stage is the model training process using the `LogisticRegression` algorithm. In this stage, the model learns from the dataset. During training, it understands patterns. It gains knowledge and uses the knowledge to make predictions.
```
Comment (Bayler) [inline content/articles/how-to-build-an-nlp-based-emotion-detection-model-using-neattext-and-scikit-learn/index.md:200]:
```suggestion
To use `Pipeline` we need to specify the machine learning stages we want to automate. In this tutorial, we have two processes we want to automate. The first stage is the `CountVectorizer` process. This stage is used to convert the raw text dataset into a matrix of numbers that a machine can understand. The second stage is the model training process using the `LogisticRegression` algorithm. In this stage, the model learns from the dataset. During training, it understands patterns, gains knowledge, and uses the knowledge to make predictions.
```

File: content/articles/how-to-build-an-nlp-based-emotion-detection-model-using-neattext-and-scikit-learn/index.md [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+### Model fitting
+To fit the pipeline stages into `x_train` and `y_train`, run this code.
+
+```python
+pipe_lr.fit(x_train,y_train)
+```
+The `Pipeline` will run the following stages automatically and produce the following output.
+
+![Pipeline ouput](/engineering-education/how-to-build-an-nlp-based-emotion-detection-model-using-neattext-and-scikit-learn/pipeline-otput.png)
+
```
Comment (Bayler) [inline content/articles/how-to-build-an-nlp-based-emotion-detection-model-using-neattext-and-scikit-learn/index.md:218]: Always proofread your work after writing. That image file name isn't correct

**Comment by Bayler** [comment]
@francis966 This is an educative tutorial. Please work on the corrections before we can move on to final review.

**Comment by francis966** [comment]
Hello @Bayler, I have made the requested changes. I have also added some more explanations to clarify some issues.
Thanks

**Comment by Bayler** [comment]
@francis966 I was going through your tutorial when I realized the word `Let's` has been overused. Try and use an alternative word/sentence instead. You can Google on alternative words to use. Once done, we can finalize this review.

**Comment by francis966** [comment]
Hello @Bayler, I have made the requested changes.

**Comment by Bayler** [comment]
@francis966 I've made some commits to improve the flow of your tutorial. Make sure to go through them to enable you improve on your next article.

**Comment by hectorkambow** [comment]
jumping on this one
